### PR TITLE
Finger-Exercise: Redux added to the tic-tac-toe game

### PR DESCRIPTION
**Pull request by alejobh**
## Summary
Finger-exercise from the card done.
Added Redux to the Tic-Tac-Toe game, redux now handles the state that previously Game Component had with two important actions: handling the player moves and handling the history changes.

## Trello Cards
https://trello.com/c/Exs9b4Mw

### CODE REVIEW

File: react/game/src/app/screens/Game/index.js [hunk 1]
```diff
@@ -1,73 +1,38 @@
 import React, { Component } from 'react';
+import { arrayOf, bool, string, number, shape, func } from 'prop-types';
+import { connect } from 'react-redux';
 
-import { winLines } from '../../../constants/game';
+import { calculateWinner } from '../../../utils/game';
+import { actionCreators } from '../../../redux/game/actions';
```
Comment (jmtov) [inline react/game/src/app/screens/Game/index.js]: I'd recommend setting up these paths in webpack so you can do absolute imports. 👍

**Review by jmtov — CHANGES_REQUESTED** [review]
Nice job! Please check out my comment. 😺

File: react/game/src/app/screens/Game/index.js (hunk 1 above)
Comment (alejobh) [reply react/game/src/app/screens/Game/index.js]: Excelent 👌🏼

File: react/game/.babelrc [hunk 2]
```diff
@@ -8,7 +8,9 @@
       "alias": {
         "@components": "./src/app/components",
         "@screens": "./src/app/screens",
-        "@redux": "./src/redux"
+        "@redux": "./src/redux",
+        "@utils": "./src/utils/",
+        "@constants": "./src/constants"
```
Comment (wfolini) [inline react/game/.babelrc:13]: 👌

File: react/game/src/app/screens/Game/index.js [hunk 3]
```diff
@@ -111,4 +75,29 @@ class Game extends Component {
...
+const mapDispatchToProps = dispatch => ({
+  playerMove: (history, winner, squares) =>
+    dispatch(actionCreators.handlePlayerMove(history, winner, squares)),
+  changeHistory: step => dispatch(actionCreators.handleHistoryChange(step))
+});
+
+export default connect(
+  mapStateToProps,
+  mapDispatchToProps
+)(Game);
```
Comment (wfolini) [inline react/game/src/app/screens/Game/index.js:85]: Let's split this component in a smart and presentational component. All redux connection should be placed in container component, and int the presentational one u must only handle the layout.
Check [this](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0) and [this](https://medium.com/@learnreact/container-components-c0e67432e005)

File: react/game/src/constants/game.js [hunk 4]
```diff
@@ -8,3 +8,14 @@ export const winLines = [
   [0, 4, 8],
   [2, 4, 6]
 ];
+
+export const initialState = {
```
Comment (wfolini) [inline react/game/src/constants/game.js]: This `initialState` must be placed in the reducer file of the store

File: react/game/src/redux/game/actions.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    const state = getState();
+    const winner = calculateWinner(state.game.history[step].squares);
+    const change = {
+      stepNumber: step,
+      xIsNext: step % 2 === 0,
+      winner
+    };
+    dispatch(privateActionCreators.gameHistoryChanged(change));
+  }
+};
```
Comment (wfolini) [inline react/game/src/redux/game/actions.js:36]: Excellent approach!

File: react/game/src/redux/game/reducer.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
+import { initialState } from '@constants/game';
+
+const reducer = (state = initialState, action) => {
```
Comment (wfolini) [inline react/game/src/redux/game/reducer.js:12]: Excellent! Don't forget move `initialState` in this file

File: react/game/src/utils/game.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import { winLines } from '@constants/game';
+
+export const concatHistory = (history, squares) => history.concat([{ squares }]);
+
+export const calculateWinner = squares => {
```
Comment (wfolini) [inline react/game/src/utils/game.js:5]: 👏 👏

File: react/game/src/app/components/App/index.js [hunk 8]
```diff
@@ -1,23 +1,14 @@
...
+
+function App() {
+  return (
+    <Provider store={store}>
+      <Game />
+    </Provider>
+  );
 }
 
 export default App;
```
Comment (wfolini) [inline react/game/src/app/components/App/index.js]: Move this file to `src/app/index.js`, `App` is a screen

**Review by wfolini — CHANGES_REQUESTED** [review]
Excellent approach! Just some comments and go!

**Review by jmtov — DISMISSED** [review]

File: react/game/src/app/screens/Game/layout.js [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+import styles from './styles.scss';
+import Board from './components/Board';
+
+function GameLayout({ history, stepNumber, checkStatus, movement, winner, xIsNext, handleClick }) {
+  const current = history[stepNumber];
+  const status = checkStatus();
+  const moves = history.map(movement);
+
+  const getClassWinner = () => {
```
Comment (wfolini) [inline react/game/src/app/screens/Game/layout.js]: I'd convert this comp to a class component, and move this func to a instance of the class, for avoid create a new instance of this function in each render

**Review by wfolini — CHANGES_REQUESTED** [review]
Last change!

File: react/game/src/app/screens/Game/layout.js (hunk 9 above)
Comment (alejobh) [reply react/game/src/app/screens/Game/layout.js]: Perfect, on my way.

**Review by wfolini — APPROVED** [review]
Excellent!

**Pull request merged**